### PR TITLE
fix: make foApprox mdata-invariant

### PR DESCRIPTION
**Pull request by Kha**

### CODE REVIEW

**Comment by Kha** [comment]
Test case by @larsk21. This came up in the context of https://github.com/larsk21/iris-lean.

File: src/Lean/Meta/ExprDefEq.lean [hunk 1]
```diff
@@ -888,14 +888,15 @@ def checkAssignment (mvarId : MVarId) (fvars : Array Expr) (v : Expr) : MetaM (O
       let v ← instantiateMVars v
       CheckAssignment.checkAssignmentAux mvarId fvars hasCtxLocals v
 
-private def processAssignmentFOApproxAux (mvar : Expr) (args : Array Expr) (v : Expr) : MetaM Bool :=
+private partial def processAssignmentFOApproxAux (mvar : Expr) (args : Array Expr) (v : Expr) : MetaM Bool :=
   match v with
+  | .mdata _ e   => processAssignmentFOApproxAux mvar args v
```
Comment (leodemoura) [inline src/Lean/Meta/ExprDefEq.lean:893]: There is typo here. `v` => `e`

Comment (leodemoura) [reply src/Lean/Meta/ExprDefEq.lean:893]: We can also fix it using 
```lean
match v.cleanupAnnotations with
```

Comment (Kha) [reply src/Lean/Meta/ExprDefEq.lean]: Ahh, that's what I get for not rechecking the test case after refactoring the code... as well as ignoring the linter warning

File: src/Lean/Meta/ExprDefEq.lean [hunk 2]
```diff
@@ -888,14 +888,15 @@ def checkAssignment (mvarId : MVarId) (fvars : Array Expr) (v : Expr) : MetaM (O
       let v ← instantiateMVars v
       CheckAssignment.checkAssignmentAux mvarId fvars hasCtxLocals v
 
-private def processAssignmentFOApproxAux (mvar : Expr) (args : Array Expr) (v : Expr) : MetaM Bool :=
+private partial def processAssignmentFOApproxAux (mvar : Expr) (args : Array Expr) (v : Expr) : MetaM Bool :=
```
Comment (leodemoura) [inline src/Lean/Meta/ExprDefEq.lean:891]: You should also be able to remove `partial` now that the typo was fixed.

Comment (Kha) [reply src/Lean/Meta/ExprDefEq.lean:891]: Right, thanks... I started out with `consumeMData` but noticed neither of them was used in other parts of the file, perhaps for performance purposes I assumed

Comment (leodemoura) [reply src/Lean/Meta/ExprDefEq.lean]: `consumeMData` was added after we wrote this file.

**Pull request merged**